### PR TITLE
UP-4837 adding removeByFName

### PR DESCRIPTION
**Pull request by davidmsibley**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [ ] tests are included
-   [x] documentation is changed or added

##### Description of change

Adds a remove by fname action to the /layout servlet

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java [hunk 1]
```diff
@@ -231,6 +232,65 @@ public ModelAndView removeElement(HttpServletRequest request, HttpServletRespons
     }
 
     /**
+     * Remove the first element with the provided fname from the layout.
+     *
+     * @param request
+     * @param response
+     * @param fname
+     * @return
+     * @throws IOException
```
Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: Could the autogenerated Javadocs lines 237-241, be removed or updated to include a description?

File: uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java [hunk 2]
```diff
@@ -231,6 +232,65 @@ public ModelAndView removeElement(HttpServletRequest request, HttpServletRespons
...
+
+            } else {
+                // Delete the requested element node.  This code is the same for
+                // all node types, so we can just have a generic action.
+                if (!ulm.deleteNode(elementId)) {
+                    logger.info(
+                            "Failed to remove element ID {} from layout root folder ID {}, delete node returned false",
+                            elementId,
+                            ulm.getRootFolderId());
+                    response.sendError(HttpServletResponse.SC_FORBIDDEN);
```
Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: If the id sent from the client cannot be handled would this be a `HttpServletResponse.SC_BAD_REQUEST` error?

Comment (davidmsibley) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: With the exception of one line, this entire function is copied from removeElement.  I agree with you that this should be a SC_BAD_REQUEST, and in uPortal 4.2 it was, however that was changed in this commit: https://github.com/Jasig/uPortal/commit/6a26a16507f116c2f907e31a1d9434f116366bec#diff-ef2d279bd054b445090ea5dc95d41d2aR210
driven by this JIRA: https://issues.jasig.org/browse/UP-4549

It's not clear to me why SC_FORBIDDEN is necessary here.

Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: @jameswennmacher @drewwills @bjagg Any insights into the changes from [UP-4549](https://issues.jasig.org/browse/UP-4549)?

Comment (bjagg) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: SC_FORBIDDEN should probably remain here. There has already been a check to verify the node exists in the layout. Suggest adding a conditional earlier to check if the elementId is null and return SC_BAD_REQUEST in that case.

File: uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java [hunk 3]
```diff
@@ -231,6 +232,66 @@ public ModelAndView removeElement(HttpServletRequest request, HttpServletRespons
...
+        ModelAndView result = new ModelAndView("jsonView", Collections.EMPTY_MAP);
+        try {
+            IUserInstance ui = userInstanceManager.getUserInstance(request);
+            IPerson per = getPerson(ui, response);
+
+            UserPreferencesManager upm = (UserPreferencesManager) ui.getPreferencesManager();
+            IUserLayoutManager ulm = upm.getUserLayoutManager();
+
+            String elementId = ulm.getUserLayout().findNodeId(new PortletSubscribeIdResolver(fname));
+            if (elementId != null) {
```
Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: Could the logical structure be [simplified with a guard clause](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html)?

File: uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java [hunk 4]
```diff
@@ -231,6 +232,66 @@ public ModelAndView removeElement(HttpServletRequest request, HttpServletRespons
...
+                boolean succeeded = true;
+                if (elementId.startsWith(Constants.FRAGMENT_ID_USER_PREFIX)
+                        && ulm.getNode(elementId) instanceof org.apereo.portal.layout.node.UserLayoutFolderDescription) {
+                    removeSubscription(per, elementId, ulm);
+                } else {
+                    // Delete the requested element node.  This code is the same for
+                    // all node types, so we can just have a generic action.
+                    succeeded = ulm.deleteNode(elementId);
+                }
+                if (succeeded) {
```
Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: Another place where a guard clause could be used?

File: uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java [hunk 5]
```diff
@@ -170,7 +171,7 @@ public void setPortletWindowRegistry(IPortletWindowRegistry portletWindowRegistr
 
     // default tab name
     protected static final String DEFAULT_TAB_NAME = "New Tab";
-
+
```
Comment (ChristianMurphy) [inline uportal-war/src/main/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServlet.java]: Could the trailing spaces be removed?

**Pull request merged**